src/visualization_integration.cu

```diff
@@ -0,0 +1,69 @@
+#include <iostream>
+#include <fstream>
+
+#include "visualization_integration.cuh"
+
+
+__host__ std::string get_visualization_endpoint()
+{
+    std::ifstream f("config/visualization_endpoint.txt", std::ios::in);
+    std::string url;
+    getline(f, url);
+    return url;
+}
+
+__host__ std::string vector_double_to_json_array(const std::vector<double> &v)
+{
+    if(v.empty())
+        return "[]";
+
+    std::string ans = "[";
+    for(double i : v)
+    {
+        ans += std::to_string(i) + ',';
+    }
+    ans.back() = ']';
+
+    return ans;
+}
+
+__host__ bool send_particles_to_visualization(const std::string &url, MapNode *nodes, int n_of_nodes)
+{
+    std::vector<double> x, y, z;
+    x.reserve(n_of_nodes);
+    y.reserve(n_of_nodes);
+    z.reserve(n_of_nodes);
+
+    for(int i = 0; i < n_of_nodes; ++i)
+    {
+        if(!nodes[i].does_contain_particle())
+            continue;
+
+        SpacePoint coords = nodes[i].get_coordinates();
+        x.push_back(coords.x);
+        y.push_back(coords.y);
+        z.push_back(coords.z);
```

Comment:
If I understand this correctly, you send coordinates of nodes. But that should be coordinates of particle, shouldn't that?

Reply:
Yes, you understand the meaning of the vectors correctly. We usually expect the coordinates of a node to not differ a lot from the coordinates of a particle located in it, but I think you are totally correct and it surely won't be worse if we use coordinates of _particles_ to tell where _particles_ are located, especially assuming that the distance between the neighbor node may vary